### PR TITLE
Unit test for peer-to-peer memory copies.

### PR DESCRIPTION
**Pull request by tdd11235813**
To test the HIP memory workarounds, I started writing a test for memory copies between two devices. If there is only one device, a boost test message is produced, but the test does not fail.

Scenario:
- allocates buffer on dev0 and dev1
- memsets buf0 on dev0 
- copies buf0 to buf1 (dev0 -> dev1)
- compares bytes set in buf1

At the moment, only `DefaultQueue` is tested, which is normally only async (doesn't exclude this some crucial test cases for alpaka?). A few data types are tested as well, which is not really needed.
I actually need to test sync + async implementation for 1D, 2D and 3D for every accelerator, where two devices are possible.
So, I think to implement another set of types for the test scenarios, just like `TestAccs` or `TestQueues`.

Would you like to have this test case in alpaka and if so, what would be the right style?

// I also checked with nvprof the transfers, see _Details_ for an example output with CUDA + Async.

<details>

```
$> CUDA_VISIBLE_DEVICES=1,2 nvprof test/unit/mem/p2p/memP2P
Running 30 test cases...
==8330== NVPROF is profiling process 8330, command: test/unit/mem/p2p/memP2P

*** No errors detected
==8330== Profiling application: test/unit/mem/p2p/memP2P
==8330== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   45.88%  34.787ms         1  34.787ms  34.787ms  34.787ms  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EElNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_lEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                   44.54%  33.770ms         1  33.770ms  33.770ms  33.770ms  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEmNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_mEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    3.14%  2.3826ms         1  2.3826ms  2.3826ms  2.3826ms  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEiNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_iEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    3.02%  2.2869ms         1  2.2869ms  2.2869ms  2.2869ms  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEjNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_jEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    1.09%  829.82us         1  829.82us  829.82us  829.82us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm2EEmNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_mEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.96%  725.98us         1  725.98us  725.98us  725.98us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm2EElNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_lEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.60%  452.19us         1  452.19us  452.19us  452.19us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm3EEtNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_tEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.22%  169.31us         1  169.31us  169.31us  169.31us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm2EEiNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_iEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.22%  166.85us         1  166.85us  166.85us  166.85us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm2EEjNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_jEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.08%  60.128us        15  4.0080us  1.5680us  16.160us  [CUDA memcpy HtoD]
                    0.07%  55.967us         1  55.967us  55.967us  55.967us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm2EEtNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_tEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.07%  53.151us        15  3.5430us  1.9190us  12.000us  [CUDA memcpy DtoH]
                    0.03%  20.672us        15  1.3780us  1.2160us  1.8240us  [CUDA memset]
                    0.02%  18.240us         1  18.240us  18.240us  18.240us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm1EElNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_lEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.02%  17.632us         1  17.632us  17.632us  17.632us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm1EEmNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_mEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.01%  10.912us         1  10.912us  10.912us  10.912us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm1EEjNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_jEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.01%  10.720us         1  10.720us  10.720us  10.720us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm1EEiNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_iEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
                    0.01%  7.4240us         1  7.4240us  7.4240us  7.4240us  _ZN6alpaka4exec4cuda6detail10cudaKernelISt17integral_constantImLm1EEtNS_4test3mem4view20VerifyBytesSetKernelEJNS8_6traits12IteratorViewIKNS_3mem3buf9BufCudaRtIjS5_tEEvEESH_hEEEvNS_3vec3VecIT_T0_EET1_DpT2_
      API calls:   89.67%  805.97ms        30  26.866ms  2.1740us  503.30ms  cudaStreamCreate
                    8.45%  75.920ms        30  2.5307ms  6.8280us  34.794ms  cudaStreamSynchronize
                    0.57%  5.1156ms        30  170.52us  156.76us  265.01us  cudaFree
                    0.26%  2.2934ms        10  229.34us  199.01us  302.02us  cudaMalloc
                    0.25%  2.2278ms        10  222.78us  197.56us  263.36us  cudaMalloc3D
                    0.25%  2.2056ms        10  220.56us  196.88us  250.04us  cudaMallocPitch
                    0.21%  1.8792ms       188  9.9950us     323ns  382.69us  cuDeviceGetAttribute
                    0.13%  1.1705ms         2  585.26us  581.50us  589.03us  cuDeviceTotalMem
                    0.05%  452.14us        60  7.5350us  3.5770us  18.529us  cudaStreamDestroy
                    0.04%  358.72us        15  23.914us  19.039us  69.143us  cudaLaunch
                    0.03%  251.57us        10  25.156us  23.503us  30.866us  cudaMemcpy3DPeerAsync
                    0.02%  179.14us         2  89.570us  84.317us  94.823us  cuDeviceGetName
                    0.02%  177.59us       180     986ns     428ns  27.606us  cudaSetDevice
                    0.01%  129.21us         5  25.842us  21.967us  38.806us  cudaMemcpyPeerAsync
                    0.01%  104.57us         5  20.913us  19.365us  25.803us  cudaMemset2DAsync
                    0.01%  100.26us         5  20.052us  19.263us  21.526us  cudaMemset3DAsync
                    0.01%  92.142us        30  3.0710us  1.6850us  12.920us  cudaStreamCreateWithFlags
                    0.01%  72.774us       360     202ns     117ns  2.6330us  cudaGetLastError
                    0.01%  53.825us         5  10.765us  8.4490us  18.918us  cudaMemsetAsync
                    0.00%  31.415us        75     418ns     126ns  10.247us  cudaSetupArgument
                    0.00%  17.985us        45     399ns     112ns  5.8570us  cudaGetDeviceCount
                    0.00%  5.2990us        15     353ns     303ns     595ns  cudaConfigureCall
                    0.00%  4.6850us         3  1.5610us     300ns  2.3930us  cuDeviceGetCount
                    0.00%  2.8810us         4     720ns     457ns     957ns  cuDeviceGet
```

### CODE REVIEW

File: test/unit/mem/p2p/src/P2P.cpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
+/**
+ * \file
+ * Copyright 2015-2017 Benjamin Worpitz
```
Comment (ax3l) [inline test/unit/mem/p2p/src/P2P.cpp]: @tdd11235813 pls add yourself :)

File: test/unit/mem/p2p/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/**
+ * \file
+ * Copyright 2015 Benjamin Worpitz
```
Comment (ax3l) [inline test/unit/mem/p2p/src/main.cpp]: @tdd11235813 pls add yourself :)

File: test/unit/mem/p2p/src/P2P.cpp (hunk 1 above)
Comment (BenjaminW3) [reply test/unit/mem/p2p/src/P2P.cpp]: The date is most probably also wrong.

**Comment by BenjaminW3** [comment]
We can simply merge this if it gets green. You could create a new compile time vector of pairs of accs and corresponding queues in a follow up PR but I could also do this.

**Comment by BenjaminW3** [comment]
You are getting some warnings with clang:
```
alpaka/test/common/include/alpaka/test/mem/view/ViewTest.hpp:57:29: error: unused function template 'testViewImmutable' [-Werror,-Wunused-template]
                static auto testViewImmutable(
                            ^
...
```
It could be due to the methods in `ViewTest.hpp` being incorrectly marked as `static`. Please try to remove this.

**Comment by tdd11235813** [comment]
I could reproduce the clang errors and fixed them by removing the `static` attributes. Nevertheless, other clang errors occurred, but seem to be module problems only. 
In travis there is another error for g++ 7.3 compiler, where undefined references occurred at memP2P target with OpenMP4 back-end. I could not reproduce it locally, although gcc7.1 was the latest one, I tested.
I also run the travis CI on a local branch, no approval needed now, just informing about the WIP.

**Comment by tdd11235813** [comment]
This single travis test fails with undefined references on ExecCpuOmp4 for offloading VerifyBytesSetKernel:
> ALPAKA_CI_DOCKER_BASE_IMAGE_NAME=ubuntu:14.04 CXX=g++     CC=gcc   ALPAKA_CI_GCC_VER=7       CMAKE_BUILD_TYPE=Release ALPAKA_CI_BOOST_BRANCH=boost-1.63.0 ALPAKA_CI_CMAKE_VER=3.8.2  ALPAKA_CI_SANITIZERS=      OMP_NUM_THREADS=3

This one succeeds:
> ALPAKA_CI_DOCKER_BASE_IMAGE_NAME=ubuntu:16.04 CXX=g++     CC=gcc   ALPAKA_CI_GCC_VER=8       CMAKE_BUILD_TYPE=Debug   ALPAKA_CI_BOOST_BRANCH=boost-1.64.0 ALPAKA_CI_CMAKE_VER=3.11.1 ALPAKA_CI_SANITIZERS=      OMP_NUM_THREADS=4

Also this one, also gcc7.3.0, Ubuntu16.04:
> ALPAKA_CI_DOCKER_BASE_IMAGE_NAME=ubuntu:16.04 CXX=g++     CC=gcc   ALPAKA_CI_GCC_VER=7       CMAKE_BUILD_TYPE=Release ALPAKA_CI_BOOST_BRANCH=boost-1.67.0 ALPAKA_CI_CMAKE_VER=3.7.2  ALPAKA_CI_SANITIZERS=      ALPAKA_ACC_GPU_CUDA_ENABLE=ON  ALPAKA_CUDA_VERSION=9.2 ALPAKA_CUDA_COMPILER=nvcc ALPAKA_CUDA_ARCH="30;35"

Compared sources memBuf - memP2P and compared compiler/linker flags, it is the same except for the actual code. Error could not reproduced with gcc7.3.1 (travis: gcc7.3.0). 
Any clues?

**Comment by tdd11235813** [comment]
This is some interesting bug, which can be seen in ubuntu14.04 with gcc versions and `-O3` or `-O2` enabled:
-  **gcc7.3.0** (ubuntu/test) and **gcc7.2.0** (jonathonf) fail to compile memP2P with `-O3` or `-O2`
- (**gcc8.1.0** (Ubuntu Core): memP2P is working.)
- (**gcc7.3.0** Taurus: memP2P is working)

When `-g` or `-O1` is used instead of `-O3`, then linker runs fine. I tried several things to get `-O3` and `-O2` working by playing around with `-foffload=disable`, `-fno-stack-protector`, `-flto`, amongst other things, but no luck ;) ([info gcc openmp](https://gcc.gnu.org/wiki/Offloading))

I also tried to produce a MWE with the same compile and linking scheme (libcommon.a and so on), however this did not produce the error: [mwe-working.zip](https://github.com/ComputationalRadiationPhysics/alpaka/files/2283224/mwe-working.zip)

I also compared the generated files. The meta files were identical, but the output files differs in size.

| | -O1 | -O2 | -O2 (gcc8.1) |
| --- | --- | --- | --- |
P2P.cpp.o | 4.6 MiB | **744 KiB** | 3.0 MiB |

So it looks like, gcc/openmp is simply not producing the machine code. 
Does it optimize code out, where it shouldn't do? But why is it working on gcc7.3.0 on Taurus?

So either gcc7 was built wrong or something, gcc7+openmp4 depends on, is wrong in the packages on the ubuntu system.
So, how to proceed? Worst case would be to compile gcc7.3 from source, if there is no other flag which prevents this false optimization.

For the logs, the undefined reference error is of the pattern ( I guess he optimized out the operator()() ):

> CMakeFiles/memP2P.dir/src/P2P.cpp.o:(.gnu.offload_funcs+0x90): Nicht definierter Verweis auf `alpaka::exec::ExecCpuOmp4<std::integral_constant<unsigned long, 1ul>, long, alpaka::test::mem::view::VerifyBytesSetKernel, alpaka::test::mem::view::traits::IteratorView<alpaka::mem::buf::BufCpu<unsigned int, std::integral_constant<unsigned long, 1ul>, long> const, void>, alpaka::test::mem::view::traits::IteratorView<alpaka::mem::buf::BufCpu<unsigned int, std::integral_constant<unsigned long, 1ul>, long> const, void>, unsigned char>::operator()() const [clone ._omp_fn.74]'

**Comment by BenjaminW3** [comment]
Does this block you locally? Building gcc by hand in CI is not really feasible (build time).
For now, you could simply exclude this code from compilation with some ifdefs. We can create a ticket and fix this strange compiler later (if ever).

**Comment by tdd11235813** [comment]
it is only a travis thing, so it does not bother me locally. I can exclude this on travis side, where a special cmake macro could be enabled, so memP2P target is not included then. The macro gets a comment which refers to an issue, I will open for it (correspondingly: memP2P + OpenMP4 + gcc7.3.0 + Ubuntu14.04 fails).

**Comment by BenjaminW3** [comment]
i would prefer to have this ifdef in source because all other such exclusions are in soirce as well. Furthermorevthis allows to keep the excluded code to a minimum.

**Comment by tdd11235813** [comment]
ok, that could work, that only memP2P code gets the ifdef guards ... by the gcc version macros and that we are running on CI via some ALPAKA_CI... macro, that is already set. Then I would not need an extra cmake code.

**Comment by tdd11235813** [comment]
ok, well, that is the current solution, which does not detect Ubuntu14.04, but only affects P2P.cpp.
(Ubuntu16.04 is actually working with gcc7.3.0).
```C
#if defined(ALPAKA_CI) &&                             \
    BOOST_COMP_GNUC >= BOOST_VERSION_NUMBER(7,2,0) && \
    BOOST_COMP_GNUC < BOOST_VERSION_NUMBER(8,0,0) && \
    defined(ALPAKA_ACC_CPU_BT_OMP4_ENABLE)
    BOOST_TEST_MESSAGE( "Currently, memP2P is not working with gcc7.2 / gcc7.3 on Ubuntu14.04 on travis/CI.");
    BOOST_CHECK(true);
#else
   // ..
#endif
```

**Pull request merged**